keps/sig-apps/3939-consider-terminating-pods-jobs/README.md

```diff
@@ -0,0 +1,712 @@
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently, Jobs and Deployments start replacement Pods as soon as previously created Pods are marked for terminating.  Terminating pods are in a transitory state where they are neither active nor really fully terminated.  
+This KEP proposes a new field for the Job controller that controls whether to count terminating
+pods as like they were active.  The goal of this KEP is to allow for opt-in behavior where terminating pods count as active.  This will allow users to see that new pods will be only created once the existing pods have fully terminated.
+
+## Motivation
+
+Existing Issues:
+
+- [Job Creates Replacement Pods as soon as Pod is marked for deletion](https://github.com/kubernetes/kubernetes/issues/115844)
+- [Kueue: Account for terminating pods when doing preemption](https://github.com/kubernetes-sigs/kueue/issues/510)
+
+Many common machine learning frameworks, such as Tensorflow and JAX, require unique pods.  If a terminating pod is not considered active, a replacement pod is created and
+can cause errors.  This is a rare case but it can provide problems if a job needs to guarantee that the existing pods terminate
+before starting new pods.  
+
+In scarce compute environments, these resources can be difficult to obtain so pods can take a long time to find resources and they may only be able to find nodes once the existing pods have been terminated.
+
+If a job is stuck in terminating, it could be possible for autoscaling to kick in and give a new node.  
+This is not ideal if you could just wait for the pod to terminating and reuse that node.  
```

Comment:
I see no need in the motivation or user stories to actually count the terminating pods as active. I see only a feature request that a pod recreation delay is required until a cleanup is performed. Is there a reason or a use case for changing the counting at the API level?

It seems to me that we are unnecessarily changing the behavior of `status.active` depending on how the `spec` is configured. We can meet the requirements by just recording the information in `status.terminating` without modifying the status behavior.


Users can have a simpler mental model for the `status.active` field as a number of pods doing useful work. The value may change depending on a number of factors, but at least the meaning would be predictable. If a user is surprised that there are no active pods, s/he can consult the status first and then go back to spec as to why this is so, but the user can always  trust the status.

Side note: we should consider the API surface first, but implementation wise this can be done by tracking activeAndTerminating pods internally, or revising each place in a job logic where active pods are considered, but we should pay attention what we write in the status.

Reply:
I thought the KEP had departed from the idea of counting terminating pods as active.
But the implementation details seem to say otherwise. I agree that we shouldn't change the concept of active.

That said, there is motivation to knowing the number of terminating pods, like story 2 states. But this is covered by having a `.status.terminating` field.

Reply:
Sorry about the confusion.  I pushed up a change.  

We are counting terminating pods separately from active.  As @alculquicondor says, there was a request that we should count the number of terminating pods.

I mention another case with @mimowo work on Retriable jobs is that we are counting active differently if `PodDisruptionCondition` is true or false.

Reply:
- I think either the implementation part needs to be a bit rewritten to be more abstract or more specific. Or/and we should clearly define the guarantees for the `.status.active` that it will not be changed.
- there is still inconsistent usage of "active" in the KEP

> I mention another case with @mimowo work on Retriable jobs is that we are counting active differently if PodDisruptionCondition is true or false.

From KEP perspective, I would not say we are counting active differently, just that there is a known bug/edge case

Reply:
We should be renaming `FilterActivePods` to `FilterActiveAndTerminatingPods` to avoid confusion. That would normally come up in the code review, but the fact that there's a snippet in the KEP makes the design confusing.

Removing that level of detail in the KEP is fine.

Reply:
I removed that section.